### PR TITLE
Serializer class

### PR DESCRIPTION
**Pull request by joakim-hove**
Slightly expand on the very basic serializer used for `SummaryState`- now also used for `UDQState`.

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

File: opm/common/utility/Serializer.hpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+
+    std::vector<char> buffer;
+private:
+    void pack(const void * ptr, std::size_t value_size) {
+        std::size_t write_pos = this->buffer.size();
+        std::size_t new_size = write_pos + value_size;
+        this->buffer.resize( new_size );
+        std::memcpy(&this->buffer[write_pos], ptr, value_size);
```
Comment (bska) [inline opm/common/utility/Serializer.hpp:86]: This statement is undefined if `value_size == 0` (i.e., when `new_size == write_pos`).  That, in turn, happens whenever someone passes an empty value such as when [serializing](https://github.com/OPM/opm-common/blob/015d8ff92397b388a5737acbc6f495f179c68782/tests/parser/UDQTests.cpp#L99-L102) the `UDQSet` named `FUPR` in the `UDQSTATE` unit test.

It is not clear to me where the actual problem is, but we must obviously never reference `buffer[buffer.size()]`.  I can sort of work around this using the patch below but this feels like a hack and I suspect the real solution is elsewhere&mdash;e.g. ensuring that we never pass `value_size == 0` into this member function.
```diff
diff --git a/opm/common/utility/Serializer.hpp b/opm/common/utility/Serializer.hpp
index 342b53ecb..795434cd6 100644
--- a/opm/common/utility/Serializer.hpp
+++ b/opm/common/utility/Serializer.hpp
@@ -80,6 +80,9 @@ public:
     std::vector<char> buffer;
 private:
     void pack(const void * ptr, std::size_t value_size) {
+        if (value_size == std::size_t{0})
+            return;
+
         std::size_t write_pos = this->buffer.size();
         std::size_t new_size = write_pos + value_size;
         this->buffer.resize( new_size );
```

Comment (joakim-hove) [reply opm/common/utility/Serializer.hpp:86]: > This statement is undefined if value_size == 0 (i.e., when new_size == write_pos)

That I agree.
 
> That, in turn, happens whenever someone passes an empty value such as when serializing the UDQSet named FUPR in the UDQSTATE unit test.

This I do not really understand; the `UDQSet` in question is not empty; but I still agree that you can create an empty set - which can create problems - will look into it.

Did you find this by staring at the code, or is it the result of some instrumented build?

Comment (bska) [reply opm/common/utility/Serializer.hpp:86]: > > This statement is undefined if `value_size == 0` (i.e., when `new_size == write_pos`)
> 
> That I agree.
> 
> > That, in turn, happens whenever someone passes an empty value such as when serializing the `UDQSet` named `FUPR` in the `UDQSTATE` unit test.
> 
> This I do not really understand; the `UDQSet` in question is not empty;

It happens when serializing the [`values`](https://github.com/OPM/opm-common/blob/015d8ff92397b388a5737acbc6f495f179c68782/src/opm/parser/eclipse/EclipseState/Schedule/UDQ/UDQSet.cpp#L568-L569) of this `UDQSet`.  At least one of the elements has an empty [`m_wgname`](https://github.com/OPM/opm-common/blob/015d8ff92397b388a5737acbc6f495f179c68782/opm/parser/eclipse/EclipseState/Schedule/UDQ/UDQSet.hpp#L59) (i.e., `m_wgname = ""`) which triggers this condition.

> Did you find this by staring at the code, or is it the result of some instrumented build?

Checked iterator build.  I got the traceback below the first time I ran the unit tests after incorporating this PR in my local tree.  Note in particular the remark concerning "out-of-bounds index 126, but ..."

---

```
Running 1 test case...
/usr/include/c++/7/debug/vector:417:
Error: attempt to subscript container with out-of-bounds index 126, but 
container only holds 126 elements.

Objects involved in the operation:
    sequence "this" @ 0x0x7ffc20beab80 {
      type = std::__debug::vector<char, std::allocator<char> >;
    }
unknown location(0): fatal error: in "UDQSTATE": signal: SIGABRT (application abort requested)
/home/bska/work/opm/src/opm.kdev/opm-common/tests/parser/UDQTests.cpp(96): last checkpoint

*** 1 failure is detected in the test module "UDQTests"
```

**Pull request merged**